src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1173,6 +1140,150 @@ ChooseGenerator::ChooseGenerator(const Node& node, GeneratorArgs generatorArgs)
+    return what.find("Unknown parser") != std::string::npos;
+}
+
+/**
+ * @private
+ * @param datetime
+ *   a datetime string in a known format or unsigned long long string.
+ * @return
+ *   The datetime as millis.
+ */
+static auto parseStringToMillis(const std::string& datetime) {
+    boost::posix_time::ptime utc_time{boost::posix_time::not_a_date_time};
+
+    auto millis = 0LL;
+    if (!datetime.empty()) {
+        for (const auto& format : formats) {
+            std::istringstream date_stream{datetime};
+            date_stream.imbue(format);
+            boost::local_time::local_date_time local_date{boost::local_time::not_a_date_time};
+            if (date_stream >> local_date) {
+                utc_time = local_date.utc_time();
+
+                millis = (utc_time - epoch).total_milliseconds();
+                break;
+            }
+        }
+    }
+    if (utc_time == boost::posix_time::not_a_date_time) {
+        try {
+            // Last gasp try to interpret unsigned long long.
+            millis = std::stoull(datetime);
+        } catch (const std::invalid_argument& _) {
+            auto msg = "^RandomDate: Invalid Dateformat '" + datetime + "'";
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax{msg});
+        }
+    }
+    return millis;
+};
+
+/**
+ * @private
+ * @param node
+ *   a node string value i.e. either a scalar or a `^String` value
+ * @return
+ *   a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<int64_t> dateStringParserGenerator(const Node& node, GeneratorArgs generatorArgs) {
+    auto generator = stringGenerator(node, generatorArgs);
+    auto datetime = generator->evaluate();
+    auto millis = parseStringToMillis(datetime);
+    return std::make_unique<ConstantAppender<int64_t>>(millis);
+}
+
+/**
+ * @private
+ * @param node
+ *   a node containing a date generator
+ * @return
+ *   a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<int64_t> dateTimeGenerator(const Node& node, GeneratorArgs generatorArgs) {
+    // Set of parsers to look when we request an int parser
+    // see int64Generator
+    const static std::map<std::string, Parser<UniqueGenerator<bsoncxx::types::b_date>>> dateParsers{
+        {"^Now",
+         [](const Node& node, GeneratorArgs generatorArgs) {
+             return std::make_unique<NowGenerator>(node, generatorArgs);
+         }},
+        {"^RandomDate",
+         [](const Node& node, GeneratorArgs generatorArgs) {
+             return std::make_unique<RandomDateGenerator>(node, generatorArgs);
+         }},
+    };
+
+    if (auto parserPair = extractKnownParser(node, generatorArgs, dateParsers)) {
+        // known parser type
+        auto generator = parserPair->first(node[parserPair->second], generatorArgs);
+        auto datetime = generator->evaluate();
+        auto millis = datetime.to_int64();
+        return std::make_unique<ConstantAppender<int64_t>>(millis);
```

Comment:
Oh, we need to be careful here. So this will evaluate only once. Whatever string you get the first time is what you have the whole time. I'm going back and forth with myself over what makes the most sense. 

Could we make a UniqueGenerator<int64_t> that wraps a UniqueGenerator<string>?

Reply:
Could you make sure to include comments with the documentation for this? That may only be the test cases, but still.

Reply:
Is the fact that  a ```{^Now: {}}``` evaluated once a surprise? 

If so then it may make sense to create a lazy generator for these two types.

Reply:
It is. I had proposed another generator of {^Once: {# Some other generator}} to allow explicitly doing that because there are use cases for it.

Reply:
ok, I'll have a look at lazy evaluating generators. To be correct / consistent all the non-constant cases should also be lazy evaluated (i.e.^RandomInt etc.... ).

There wil be a side effect though that the min / max check could pass or fail depending on what values are generated.